### PR TITLE
pack-package-refactoring

### PR DESCRIPTION
**Pull request by dwillist**
Signed-off-by: dwillist <dthornton@vmware.com>

### CODE REVIEW

File: text/0000-pack-package-refactor.md
Comment (ekcasey) [inline text/0000-pack-package-refactor.md:44]: Given that whatever would live here could only be used by the `stack` package, should we just include it in the `stack` package but not export it?

Also, this package would be named `internal` which is a bit strange. If we really need an internal package (the logic is used by multiple exported packages) maybe we want `pkg/internal/X` where X describes the whatever utility it is exposing. I think it would be easier to maintain a useable and well organized public interface for `stacks` etc. if any stack specific functions had to go in the external package and internal packages were reserved for coherent but non-domain-level concepts (e.g. `blob`)

File: text/0000-pack-package-refactor.md
Comment (ekcasey) [inline text/0000-pack-package-refactor.md:36]: This seems like a brand new feature rather than a refactor. Maybe this should be out of scope.

File: text/0000-pack-package-refactor.md
Comment (ekcasey) [inline text/0000-pack-package-refactor.md:49]: Given a lifecycle `prepare` phase may eventually want to use some of this logic (resolving buildpack uris, pulling buildpacks) and we don't want the lifecycle to import pack, should we think about creating a new repo (`distribution`?) for these packages?

We could do this later but it might make sense to do know, while we are refactoring to avoid problems disentangling later (if for example internal packages are used by packages that should be moved and also by packages that should remain in this repo).

File: text/0000-pack-package-refactor.md
Comment (jromero) [reply text/0000-pack-package-refactor.md:36]: It is my understanding that this structures is based on some [future thinking hypotheticals](https://github.com/buildpacks/community/discussions/51#discussion-1608992). I'd propose keeping it in the RFC but making note that it's "future thinking".

```suggestion
│   └── fetcher/ → cli to download buildpacks (future thinking)
```

File: text/0000-pack-package-refactor.md
Comment (jromero) [reply text/0000-pack-package-refactor.md:49]: Part of me is hesitant to move stuff into a new repository at this moment because we don't yet have the right seams and am afraid that the effort would prevent us from moving forward to even a slightly better structure.

In my mind, this RFC can be seen as the first step into potentially moving stuff out. I would think of it as a "promotion" strategy for any logic that exists in pack that we may want to redistribute.

ie:
`/pack/archive` -> `imgutil/archive`
`/pack/pkg/resolve` -> `<TBD>/resolve`

Comment (ekcasey) [reply text/0000-pack-package-refactor.md:49]: The makes sense. We can promote as a first step and think about extraction later if it is warranted.

File: text/0000-pack-package-refactor.md
Comment (jromero) [inline text/0000-pack-package-refactor.md:20]:
```suggestion
- Why should we do this?
    - Maintaining a modular and re-usable codebase enables contributors to be more effective, and encourages new adopters. 
    - Allow for better consumption of pack as a library by having more isolated and focused packages.
    - Identify and create logical seams that can be used to then more easily redistribute packages.
```

File: text/0000-pack-package-refactor.md
Comment (jromero) [inline text/0000-pack-package-refactor.md:57]: Not sure I understand the question and thereby the answer. Could you elaborate?

File: text/0000-pack-package-refactor.md
Comment (jromero) [inline text/0000-pack-package-refactor.md:3]:
```suggestion
- Name: Refactor Pack go packages
```

File: text/0000-pack-package-refactor.md
Comment (jromero) [reply text/0000-pack-package-refactor.md:38]: This has [already started](https://github.com/buildpacks/imgutil/pull/73) but I wouldn't say it's a blocker for this RFC.

File: text/0000-pack-package-refactor.md
Comment (ekcasey) [reply text/0000-pack-package-refactor.md:36]: Personally I would rather cut it from this RFC. I have misgivings about breaking `pack` into smaller binaries and believe worthy of its own RFC.

The `(future thinking)` caveat makes me feel slightly better, but also I don't see a reason to include this at all. The question of where we would put future CLIs doesn't seem like a controversy that need resolution (they always go in `cmd` by convention).

File: text/0000-pack-package-refactor.md
Comment (dwillist) [reply text/0000-pack-package-refactor.md:57]: The question is from the generic `0000-template.md`. The general thinking was this entire RFC seems like we are proposing a top down view of how we would like things to look. When actually performing the refactor we might decide that adjustments should be made. I believe it would be prudent to keep this RFC open and adjust it as we do the refactoring work.

**Comment by jromero** [comment]
Based on our conversionation during the sub-team sync, I believe the next steps for this RFC would be to:

1. Focus goal on the intention as opposed to the end result.
2. Provide an end result for what the expected outcome may look like as soon as the RFC is approved.
3. Provide an end result for the _potential_ long-term vision of what the structure may be.

File: text/0000-pack-package-refactor.md
Comment (dwillist) [reply text/0000-pack-package-refactor.md]: I think re-framing the goals of this rfc as focusing on a process to pull functionality out means we can defer this conversation until we need to distribute additional CLIs

**Comment by dwillist** [comment]
Hey maintainers any feedback would be appreciated @jromero @dfreilich

**Comment by jkutner** [comment]
@dwillist can you add `.idea/` to `.gitignore`. It's ok to have it in this PR IMO

File: text/0000-pack-package-refactor.md
Comment (jkutner) [inline text/0000-pack-package-refactor.md:2]: Are there any alternative proposals for the structure?

File: text/0000-pack-package-refactor.md
Comment (jkutner) [inline text/0000-pack-package-refactor.md:38]: everything below here feels like "How it works". Can you add that header?

File: text/0000-pack-package-refactor.md
Comment (jkutner) [inline text/0000-pack-package-refactor.md:91]: Are there any examples of other projects that have a similar structure to this?

Comment (jkutner) [reply text/0000-pack-package-refactor.md:91]: internal to `buildpacks` or external

File: text/0000-pack-package-refactor.md
Comment (dwillist) [reply text/0000-pack-package-refactor.md:2]: Added 👍

**Review by jromero — APPROVED** [review]
🚀 I'm all for this refactor. As I work on Tekton even more, there are certain parts of `pack` that I'd like to encapsulate and externalize to be reused on the Tekton pipeline. Simply being able to gain access to them would greatly reduce what otherwise may be duplicate efforts.

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:21]:
```suggestion
    - General refactoring benefits: 
```

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:30]:
```suggestion
1) Extract code from the `interal` package to `/pkg/<pkg-name>/...` to enable external consumption.
```
etc all the way down, to ease in case of reordering

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:66]:
```suggestion
Further, the structure itself is very intimidating to a new-comer - what is important? Where does the definition/implementation live, and why is it split into multiple components without clear organization?
```

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:57]:
```suggestion
```

These don't exist by default, they are just created through the testing structure.

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:80]: Should `internal` be inside the `pkg` dir, or a top-level package? I'm not sure what best/typical practices here are

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:106]: Assumedly part of `imgutil`, per the discussion here: https://github.com/buildpacks/imgutil/pull/73

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:112]:
```suggestion
- A more or less ambitious short term target.
```

File: text/0000-pack-package-refactor.md
Comment (dfreilich) [inline text/0000-pack-package-refactor.md:25]: Add a reference to the discussion we had here?  https://github.com/buildpacks/community/discussions/51

**Review by dfreilich — APPROVED** [review]
overall, definitely happy to see this!

File: text/0000-pack-package-refactor.md
Comment (dwillist) [reply text/0000-pack-package-refactor.md:80]: [This](https://github.com/golang-standards/project-layout) seems to say top level is expected.

File: text/0000-pack-package-refactor.md
Comment (dwillist) [reply text/0000-pack-package-refactor.md:64]: Thanks for this added 👍

**Comment by jromero** [comment]
TODO: @dwillist, remove binaries from "future" possibility.

File: text/0000-pack-package-refactor.md
Comment (jromero) [reply text/0000-pack-package-refactor.md:93]: See Prior Art section

**Review by hone — APPROVED** [review]

**Pull request merged**
